### ISSUE TITLE
Header guard not working for EM_JS functions - duplicate symbol

### ISSUE THREAD

**Issue by DolphinIQ** [emscripten-core/emscripten#21364]
Please include the following in your bug report:

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.50 (047b82506d6b471873300a5e4d1e690420b582d0)
clang version 18.0.0
Target: wasm32-unknown-emscripten
Thread model: posix
```
Windows 11 (no judgement please :)

**Failing command line in full:**
`em++ -I ./src/cpp src/cpp/*.cpp -Wall -g3 -o src/wasm/main.js`

I have defined a couple of JS functions I need inside my C++/wasm project. All worked fine when I run things in just one `.cpp` file. However it was getting too large and I split it up. ATM just 3 header files and 2 cpp files. Both cpp files need access to those JS functions, so I put their declarations into a `js_functions.h` file:
```C++
#pragma once
#include <emscripten.h>

EM_JS( void, log, ( float num ), {} );
EM_JS( void, drawSquares, ( u16 num ), {} );
EM_JS( void, drawImgIndex, ( u8 imgIndex, float x, float y, float angle ), {} );
EM_JS( float, random_f, (), {} );
```
and then included it in both cpp files. However then I started getting duplicate symbol linker errors, one such block for each function defined with `EM_JS`:
```
wasm-ld: error: duplicate symbol: __em_js__log
>>> defined in C:\pathtofiles\file_0.o
>>> defined in C:\pathtofiles\main_1.o
```
The problem goes away if I include the JS functions header in only one cpp file and remove its contents from the other cpp file. This makes me think the `EM_JS` declarations are parsed twice despite me having a header guard (tried both `#pragma once` and `ifndef`). Of course I need them in both cpp files and likely in even more in the future so its a big problem for me.

Is this a bug or am i doing something wrong? I know a bit of C and C++ but im primarily a JS dev, so I'd appreciate any help. Thanks 😊

**Comment by sbc100** [comment]
`#pragma` just mean that the header will only only be included once per compilation unit.    If you define symbol in the header they will still conflict with other compilation units that include the same header.

In this case its better to have just one of the `.cpp` files define the `EM_JS` functions and then simply declare them in the header like any other C function.

**Comment by DolphinIQ** [comment]
Hey, thanks for clarification. I moved the `EM_JS` definitions to `main.cpp` and left only declarations in `js_functions.h` 
```C++
#pragma once
void log( float num );
void drawSquares( u16 num );
void drawImgIndex( u8 imgIndex, float x, float y, float angle );
float random_f();
```
However it seems like it doesnt see the definitions now. I unfortunately got an error from the second cpp file which uses those functions:
```
wasm-ld: error: C:\pathtofiles\file_0.o: undefined symbol: random_f()
```

**Comment by sbc100** [comment]
Ah yes, I think calling EM_JS functions from other objects might have some issues still.

Can you try writing a normal C function called `random_f` that just calls the EM_JS function (perhaps call it `random_f_js`)?

**Comment by DolphinIQ** [comment]
Wrappers do work, thanks! 
However I was also able to solve it without them, with just bare `EM_JS` definitions. It turns out the setup you suggested in the first reply and one I implemented in my previous comment was correct, but in the `js_functions.h` header, I just needed to wrap the function declarations in `extern "C"` like so:
```C++
extern "C" {
    void log( float num );
    void drawSquares( u16 num );
    void drawImgIndex( u8 imgIndex, float x, float y, float angle );
    float random_f();
}
```

I hate C++ sometimes... -_-
Thank you for help!
